sematic/plugins/abstract_metrics_storage.py

```diff
@@ -0,0 +1,204 @@
+"""
+Module declaring the interface for metrics storage plug-ins.
+"""
+# Standard Library
+import abc
+import enum
+from dataclasses import dataclass, field
+from datetime import datetime
+from typing import List, Optional, Sequence, Tuple, Type, cast
+
+# Sematic
+from sematic.abstract_plugin import AbstractPlugin, PluginScope
+from sematic.config.settings import get_active_plugins
+from sematic.metrics.metric_point import MetricPoint, MetricsLabels
+
+
+@dataclass
+class MetricsFilter:
+    """
+    Data structure to query metrics.
+    """
+
+    name: str
+    from_time: datetime
+    to_time: datetime
+    labels: MetricsLabels
+
+
+@dataclass
+class MetricSeries:
+    """
+    Data structure to represent a metric series for queries to return.
+
+    Series are typically aggregated by labels.
+
+    Parameters
+    ----------
+    metric_name: str
+        The name of the metric
+    metric_type: str
+        One of `MetricType`. Using the str value instead of Enum instance for
+        serializability.
+    series: List[Tuple[float, Tuple[str, ...]]]
+        The metric series. A list of tuples. Every element in the list
+        correspond to a series value. The tuple contains two elements: the
+        value, and a tuple of label values. The order of label values
+        corresponds to the order of label names in group_by_labels.
+    group_by_labels: List[str]
+        The list of label names the metric was aggregated over.
+
+    Examples
+    --------
+    Run count with no aggregations.
+    ```
+    MetricSeries(
+        metric_name="sematic.func_run_count",
+        metric_type="COUNT",
+        series=[(483.0, ())],
+        group_by_labels=[]
+    )
+    ```
+
+    Success rate by calculator path by day
+    ```
+    MetricSeries(
+        metric_name="sematic.func_success_rate",
+        metric_type="GAUGE",
+        series=[
+            (0.68, ("path.to.foo", "2023-04-12")),
+            (0.95, ("path.to.foo", "2023-04-11")),
+            (0.45, ("path.to.bar", "2023-04-12")),
+            ...
+        ],
+        group_by_labels=["calculator_path", "date"],
```

Comment:
Treating the date/time as "just another label" is a neat trick that works fine for SQL-based plugin implementations, but I don't know that it will work out well for other dedicated metric storage plugins (like postgres). For them, time is a first-class-citizen that is handled in queries differently from labels.

Reply:
Date is NOT treated as "just another label". We do not persist the metric_time as label, we persist as a proper `TIMESTAMP` field, and that is how we are able to aggregate on `date` using SQL's native date extracting logic.
In `group_by_labels`, "labels" does not refer to metric point labels. It refers to the labels of the values in the second element of the series tuples. We can rename to `group_by_fields` if you prefer, or simply `group_by`.